### DISCUSSION TITLE
2048

### DISCUSSION THREAD

**Discussion by peterhellberg** [aduros/wasm4#101] [General]
I've started working on a little **2048** clone for WASM-4 using [TinyGo](https://tinygo.org/) this weekend. I initially started by porting a JavaScript implementation of **2048** but it turned out to be too much of a hassle 😂  

Instead I've now based the initial logic on <https://github.com/chhabraamit/2048>

(While looking around for a minimal implementation of **2048** I also found <https://github.com/hirschsn/mini2048> which nicely compiled using `CC="zig cc" make`)

The palettes are copied from the [Lospec Palette List](https://lospec.com/palette-list)

> There are still a lot of things to fix before the game can be considered done

### 🎮  [Play](https://peter.tilde.team/games/w4-2048/) (Arrow keys, X, Z)
 - 🎵  [Play with music](https://peter.tilde.team/games/w4-2048/with-music.html)
 - 🔉🎵  [Play with sound and music](https://peter.tilde.team/games/w4-2048/with-sound-and-music.html)

#### Initial output
![https://i.imgur.com/8cwt9AM.png](https://i.imgur.com/8cwt9AM.png)

#### After a few small tweaks
![https://i.imgur.com/OjmSzwR.png](https://i.imgur.com/OjmSzwR.png)

#### A few palettes from Lospec

![https://i.imgur.com/UUKenr8.png](https://i.imgur.com/UUKenr8.png)
![https://i.imgur.com/9YMMbP1.png](https://i.imgur.com/9YMMbP1.png)
![https://i.imgur.com/An3l0Wi.png](https://i.imgur.com/An3l0Wi.png)
![https://i.imgur.com/hqYu8de.png](https://i.imgur.com/hqYu8de.png)
![https://i.imgur.com/KdpatZL.png](https://i.imgur.com/KdpatZL.png)

**Comment by aduros** [comment]
Wow, this is awesome! I'm slightly addicted to 2048, so this is getting saved onto my phone :smile: Lovely graphics and palette choices.

When it's ready you should totally submit it to [GBJam](https://itch.io/jam/gbjam-9) (which is happening right now!)

**Reply by peterhellberg** [reply]
The [palettes](https://github.com/peterhellberg/w4-2048/blob/master/palettes/palettes.go) in questions are:
```go
package palettes

type Palette [4]uint32

var (
	BluesGB     = Palette{0xe5f1f3, 0x7ba8b8, 0x30617b, 0x08263b} // https://lospec.com/palette-list/blues-gb (reversed)
	EN4         = Palette{0xfbf7f3, 0xe5b083, 0x426e5d, 0x20283d} // https://lospec.com/palette-list/en4
	GBChocolate = Palette{0xffe4c2, 0xdca456, 0xa9604c, 0x422936} // https://lospec.com/palette-list/gb-chocolate
	Grapefruit  = Palette{0xfff5dd, 0xf4b26b, 0xb76591, 0x65296c} // https://lospec.com/palette-list/grapefruit (reversed)
	GreyMist    = Palette{0xf1ffe0, 0x988171, 0x463534, 0x1e1721} // https://lospec.com/palette-list/grey-mist
	IceCreamGB  = Palette{0xfff6d3, 0xf9a875, 0xeb6b6f, 0x7c3f58} // https://lospec.com/palette-list/ice-cream-gb
	Keeby       = Palette{0xc5ccb8, 0x899b98, 0x5d6872, 0x333343} // https://lospec.com/palette-list/keeby
	Lightgreen  = Palette{0xf4fbd0, 0x68cf68, 0x1e9178, 0x05241f} // https://lospec.com/palette-list/lightgreen
	Platinum    = Palette{0xe0f0e8, 0xa8c0b0, 0x507868, 0x183030} // https://lospec.com/palette-list/platinum
	Warmlight   = Palette{0xffd191, 0xff924f, 0x66605c, 0x211e20} // https://lospec.com/palette-list/warmlight (reversed)

	All = []Palette{
		BluesGB,
		EN4,
		GBChocolate,
		Grapefruit,
		GreyMist,
		IceCreamGB,
		Keeby,
		Lightgreen,
		Platinum,
		Warmlight,
	}
)
```

- https://lospec.com/palette-list/blues-gb (reversed)  
- https://lospec.com/palette-list/en4                  
- https://lospec.com/palette-list/gb-chocolate         
- https://lospec.com/palette-list/grapefruit (reversed)
- https://lospec.com/palette-list/grey-mist            
- https://lospec.com/palette-list/ice-cream-gb         
- https://lospec.com/palette-list/keeby                
- https://lospec.com/palette-list/lightgreen           
- https://lospec.com/palette-list/platinum             
- https://lospec.com/palette-list/warmlight (reversed)

**Comment by peterhellberg** [comment]
## 🎵  [Music](https://peter.tilde.team/games/w4-2048/with-music.html)
## 🔉🎵   [Sound+Music](https://peter.tilde.team/games/w4-2048/with-sound-and-music.html)

Some initial experiments with adding sound to my game. Faking timekeeping by counting rendered frames. 🤔
